### ISSUE TITLE
Can we support pickled Python code instead of uploading scripts separately? 

### ISSUE THREAD

**Issue by jdries** [Open-EO/openeo-api-poc#5]
Using the cloudpickle module, we can pickle a python lamba or function on the client side, and send it to the backend for execution. This simplifies things for the user, as he doesn't have to explicitly create processes by uploading scripts.
Is this something we can support?
Not sure to what extent Python version in client and backend have to be aligned in order for this to work.
My current client implementation will be at least Python 3.5.

**Comment by jdries** [comment]
@appelmar 
I've implemented a poc, (client and server) that shows how this can be implemented. This is an example of a REST service call, that computes a point timeseries after applying a custom function. The function is first pickled (using cloudpickle) and then base64 encoded to be able to send it as json, code can be found in my repos:
Handling request:
 <Request 'http://127.0.0.1:5000/v0.1/timeseries/point?x=1.0&y=-3.0&srs=EPSG:4326' [POST]>
Post data: 
`{
   "process_id":"band_arithmetic",
   "args":{
      "imagery":{
         "product_id":"SENTINEL2_RADIOMETRY"
      },
      "bands":[
         "B0",
         "B1",
         "B2"
      ],
      "function":"gASVZgEAAAAAAACMF2Nsb3VkcGlja2xlLmNsb3VkcGlja2xllIwOX2ZpbGxfZnVuY3Rpb26Uk5QoaACMD19tYWtlX3NrZWxfZnVuY5STlGgAjA1fYnVpbHRpbl90eXBllJOUjAhDb2RlVHlwZZSFlFKUKEsCSwBLAksDS0NDEHwAZAEZAHwAZAIZABcAUwCUTksASwGHlCmMBWNlbGxzlIwGbm9kYXRhlIaUjEcvaG9tZS9kcmllc2ovcHl0aG9ud29ya3NwYWNlL29wZW5lby1nZW90cmVsbGlzLWludGVncmF0aW9udGVzdHMvdGVzdC5weZSMCDxsYW1iZGE+lEsIQwCUKSl0lFKUSv////99lIeUUpR9lCiMB2dsb2JhbHOUfZSMCGRlZmF1bHRzlE6MBGRpY3SUfZSMBm1vZHVsZZSMCF9fbWFpbl9flIwOY2xvc3VyZV92YWx1ZXOUTowIcXVhbG5hbWWUaBB1dFIu"
   }
}
`

**Comment by edzer** [comment]
@jdries what would be the benefit of doing this, over plain text scripts?

**Comment by jdries** [comment]
@edzer There's two separate improvements here:

- Including a script directly in the POST call makes the service entirely stateless. (As opposed to uploading a script in a separate call, which would be statefull). Stateless web services are (a lot) easier to implement.
- Pickling the lambda allows me to provide a very simple client API, where the user does not really see that his function is remotely executed. Have a look at this snippet (from :https://github.com/Open-EO/openeo-python-client/blob/master/tests/test_bandmath.py:)
`
#define a computation to perform
bandFunction = lambda band1, band2, band3: band1 + band2
ndvi_coverage = s2_radio.combinebands([s2_radio.bands[0],s2_radio.bands[1],s2_radio.bands[2]], bandFunction)
`
I got this idea from pyspark. 
Note that we might also be able to support inserting a plain text script here, which might be useful for instance when calling our API from a javascript client which does not know how to deal with python code.

**Comment by m-mohr** [comment]
I support that having a stateless API is better, but it might be a problem when a user wants to upload multiple/big files and run scripts multiple times. This might lead to additional costs (see AWS cost structure for IO) and requires more time and bandwidth in case files are uploaded with each call again. The question is whether we want to support only one way or both?

Additionally, we shouldn't have programming language specifics in the API that only suits to one language, e.g. Python. User defined function in the API allow to specify a language. In case we decide to implement this suggestion, we should add the language and version to the request, too. Therefore, it is not restricted to Python only.

In my opinion it should be more like this example:
```javascript
{
   "process_id":"band_arithmetic",
   "args":{
      "imagery":{
         "product_id":"SENTINEL2_RADIOMETRY"
      },
      "bands":[
         "B0",
         "B1",
         "B2"
      ],
      "udf":{
         "language":"Python",
         "version":"3.5",
         "function":"gASVZgEAAAAAAACMF2Nsb3VkcGlja2xlLmNsb3VkcGlja2xllIwOX2ZpbGxfZnVuY3Rpb26Uk5QoaACMD19tYWtlX3NrZWxfZnVuY5STlGgAjA1fYnVpbHRpbl90eXBllJOUjAhDb2RlVHlwZZSFlFKUKEsCSwBLAksDS0NDEHwAZAEZAHwAZAIZABcAUwCUTksASwGHlCmMBWNlbGxzlIwGbm9kYXRhlIaUjEcvaG9tZS9kcmllc2ovcHl0aG9ud29ya3NwYWNlL29wZW5lby1nZW90cmVsbGlzLWludGVncmF0aW9udGVzdHMvdGVzdC5weZSMCDxsYW1iZGE+lEsIQwCUKSl0lFKUSv////99lIeUUpR9lCiMB2dsb2JhbHOUfZSMCGRlZmF1bHRzlE6MBGRpY3SUfZSMBm1vZHVsZZSMCF9fbWFpbl9flIwOY2xvc3VyZV92YWx1ZXOUTowIcXVhbG5hbWWUaBB1dFIu"
      }
   }
}
```

**Comment by jdries** [comment]
I certainly agree with your updated example! As for avoiding IO costs, wouldn't viewing and download of EO imagery be the bottleneck there rather than process graphs containing bits of code?
Of course, we could also imagine providing support for uploading packaged dependencies, so that a user can import those as well in their process graphs.

**Comment by m-mohr** [comment]
Yes, you are probably right. I was thinking about uploading packaged dependencies or custom data that might be needed for computations, e.g. data to compare against or whatever. But that is probably a case for async calls and not for synchronous calls. So no problem there...
